### PR TITLE
Fix OpenMP Launch Mem Leak and Add Clang Sanitizer Build

### PR DESCRIPTION
**Pull request by rchen20**
# Summary

- This PR is a bugfix
- It does the following:
  - Fixes memory leak in the Launch OpenMP backend. Memory leak was found via clang ASAN.
  - Updates some Intel build scripts.
  - Adds build script for clang's ASAN and UBSAN sanitizers on TOSS4. 

- Note: Does _not_ fix the OpenMP Launch reduction tests. Will iterate with LC on those.

### CODE REVIEW

File: host-configs/lc-builds/toss4/clang_X_asan.cmake [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# and RAJA project contributors. See the RAJA/LICENSE file for details.
+#
+# SPDX-License-Identifier: (BSD-3-Clause)
+###############################################################################
+
+set(RAJA_COMPILER "RAJA_COMPILER_CLANG" CACHE STRING "")
+
+set(CMAKE_CXX_FLAGS_RELEASE "--gcc-toolchain=/usr/tce/packages/gcc/gcc-10.3.1 -O3 -march=native -funroll-loops -finline-functions" CACHE STRING "")
+set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "--gcc-toolchain=/usr/tce/packages/gcc/gcc-10.3.1 -O3 -g -march=native -funroll-loops -finline-functions" CACHE STRING "")
+set(CMAKE_CXX_FLAGS_DEBUG "--gcc-toolchain=/usr/tce/packages/gcc/gcc-10.3.1 -O0 -g -fsanitize=address -fno-omit-frame-pointer" CACHE STRING "")
```
Comment (MrBurmark) [inline host-configs/lc-builds/toss4/clang_X_asan.cmake:12]: You can enable sanitizers with any level of optimization, though if there are no symbols the backtraces aren't very useful

Comment (rchen20) [reply host-configs/lc-builds/toss4/clang_X_asan.cmake]: Good call, just updated the other optimization levels with the sanitizer flags. However, the build script itself sets `Debug` as the CMake build type, so it should automatically use the `-O0 -g` settings.

**Review by MrBurmark — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]
I we no longer have an asan build and test in our GitLab CI, we should add it.

File: include/RAJA/policy/openmp/launch.hpp [hunk 2]
```diff
@@ -77,6 +77,9 @@ struct LaunchExecute<RAJA::omp_launch_t> {
       ctx.shared_mem_ptr = (char*) malloc(launch_params.shared_mem_size);
 
       expt::invoke_body(f_params, loop_body.get_priv(), ctx);
+
+      free(ctx.shared_mem_ptr);
+      ctx.shared_mem_ptr = nullptr;
```
Comment (artv3) [inline include/RAJA/policy/openmp/launch.hpp:82]: Thank you for finding this!

**Review by artv3 — APPROVED** [review]

**Pull request merged**